kliff/models/parameter.py

```diff
@@ -1,705 +1,390 @@
-            # parameters by this_class.read(filename="model_params.txt")
-
-            A
-            DEFAULT
-            1.1
-
-            B
-            DEFAULT FIX
-            1.1     FIX
-
-            C
-            DEFAULT  0.1  INF
-            1.0      INF  2.1
-            2.0      FIX
-        """
-
-        with open(filename, "r") as fin:
-            lines = fin.readlines()
-            lines = _remove_comments(lines)
-
-        num_line = 0
-        while num_line < len(lines):
-            name = lines[num_line].strip()
-            num_line += 1
-
-            if name in self._params:
-                warnings.warn(
-                    f"file `{filename}`, line {num_line}. Parameter `{name}` already "
-                    f"set. Reset it.",
-                    category=Warning,
-                )
-
-            if name not in self.model_params:
-                raise ParameterError(
-                    f"file `{filename}`, line {num_line}. Parameter `{name}` not "
-                    f"supported."
-                )
-
-            num_components = len(self.model_params[name])
-
-            settings = []
-            for j in range(num_components):
-                settings.append(lines[num_line].split())
-                num_line += 1
-
-            self.set_one(name, settings)
-
-    def set(self, **kwargs):
-        """
-        Set the parameters that will be optimized.
-
-        One or more parameters can be set. Each argument is for one parameter, where the
-        argument name is the parameter name, the value of the argument is the
-        settings(including initial value, fix flag, lower bound, and upper bound).
-
-        The value of the argument should be a list of list, where each inner list is for
-        one component of the parameter, which can contain 1, 2, or 3 elements.  See
-        `~kliff.model.parameter.OptimizingParameters.read()` for the options of the
-        elements.
-
-        Example:
-            instance.set(A=[['DEFAULT'], [2.0, 1.0, 3.0]], B=[[1.0, 'FIX'], [2.0, 'INF', 3.0]])
+            arr: Array to copy to self.
         """
-        for name, settings in kwargs.items():
-            if name in self._params:
-                msg = f"Parameter `{name}` already set. Reset it."
-                warnings.warn(msg, category=Warning)
+        # transform the array and ensure that the parameter is transformed
+        arr = self.transform(arr)
+        if not self._is_transformed:
+            self.transform_()
+        self.copy_to_param(arr)
 
-            if name not in self.model_params:
-                raise ParameterError(f"Parameter `{name}` not supported.")
+    def get_numpy_array(self) -> np.ndarray:
+        """Get a numpy array of parameters in the original space.
 
-            self.set_one(name, settings)
+        This method should be uses for getting the numpy array of parameters where the ``Parameters`` class might not work.
```

Comment:
This is to ensure that we can get exact parameters back in correct space as the model expect. Since original parameters can be in either space, the model only needs to concern itself with the array that this method provides. 

I am not sure why I thought that we cannot pass it to the optimizer, as that was the explicit reason for this redesign. Although for optimizer, one should use `get_opt_numpy_array` function as it also applies masks etc. But your point stands correct regardless. I have removed the Optimizer part from it for now, unless I can remember any specific reason for it.